### ISSUE TITLE
<link> elements in <Head> break FontAwesome(!)

### ISSUE THREAD

**Issue by philiparvidsson** [vercel/next.js#20682]
# Bug report

## Describe the bug

As much as the title might sound almost facetious - it is not. It's a very difficult bug for me to figure out, but it seems to be related to `next-head-count` or something about how Next.js is parsing tags inside <Head>. I have tried to reproduce it with `create-react-app` but cannot. The bug occurs in production builds as well.

The only difference I can see with <link> vs no <link> is that the `.svg-inline--fa` class (inside <style> element, it seems) is never loaded, and that `next-head-count` is changing.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Create Next.js app
2. Install packages: `npm i --save @fortawesome/fontawesome-svg-core @fortawesome/free-solid-svg-icons @fortawesome/react-fontawesome`
3. Modify _app.tsx:

```typescript
import Head from "next/head";
import { faUser } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import "../styles/globals.css";

function MyApp({ Component, pageProps }) {
  return (
    <>
      <Head>
        <link rel="foo" href="bar" />
      </Head>
      <div>
        hello world <FontAwesomeIcon icon={faUser} />
      </div>
    </>
  );
}

export default MyApp;
```

## Expected behavior

The icon is shown as it should be.

## Actual behavior

The icon is huge.

## Screenshots

![image](https://user-images.githubusercontent.com/1163299/103452616-78977600-4cd1-11eb-9237-967dff9d25e2.png)

## System information

- OS: Windows 10 20H2 19042.685 x64
- Browser: Edge Chromium
- Version of Next.js: 10.0.4
- Version of Node.js: v15.5.0
- Deployment: all

## Additional context

You can probably guess what happens when I try to add `<link rel="apple-touch-icon" href="/images/app-icon.png" />` to my web apps...

**Comment by jamesmosier** [comment]
You'll need to import the Font Awesome CSS explicitly in `_app.js`: `import '@fortawesome/fontawesome-svg-core/styles.css';`

**Comment by philiparvidsson** [comment]
> You'll need to import the Font Awesome CSS explicitly in `_app.js`: `import '@fortawesome/fontawesome-svg-core/styles.css';`
> 
> 
> 
> 

I understand, but the question is still why there is a difference when a <link> tag is included?!

**Comment by philiparvidsson** [comment]
@jamesmosier this seems to definitely be a bug in Next.js.

Compare left side (with &lt;link&gt;) and right side (without &lt;link&gt;):

![image](https://user-images.githubusercontent.com/1163299/103622347-e55f7a00-4f36-11eb-8ace-04809f7370cf.png)

It's obvious that &lt;style&gt; elements are not loaded properly when I insert the &lt;link&gt;-tag.

**Comment by philiparvidsson** [comment]
I'm experimenting a bit more. The entire <style> block from Font Awesome is undone when a `<link>` tag is inserted.

Consider this (broken):

```javascript
  return (
    <>
      <Head>
        <link />
      </Head>
      <div>
        hello world <FontAwesomeIcon icon={faUser} />
      </div>
    </>
  );
```

And compare it to this (working):

``` javascript
  return (
    <>
      <Head>
        <meta />
      </Head>
      <div>
        hello world <FontAwesomeIcon icon={faUser} />
      </div>
    </>
  );
```

Notice how I only changed `link` to `meta`. You can use whatever, but as soon as you change it to `link` it breaks and the `style` element disappears.

**Comment by philiparvidsson** [comment]
So after some research, it's evident that it's related to `next-head-count` (see https://github.com/vercel/next.js/pull/8020/).

I don't know enough to pinpoint it, but it's the `updateHead` function that is breaking something by not updating the head properly. Possible, it's because the `next-head-count` is set as the number of children rendered (tags inserted by scripts being run are not accounted for).

**Comment by markitics** [comment]
> You'll need to import the Font Awesome CSS explicitly in `_app.js`: `import '@fortawesome/fontawesome-svg-core/styles.css';`

Thank you! I did NOT see that in the [Font Awesome React Documentation](https://fontawesome.com/how-to-use/on-the-web/using-with/react), but it's made the difference for me. `<link rel="shortcut icon" href="…" />` in `_app.js` had previously been breaking FontAwesome, but it's all good now.

**Comment by philiparvidsson** [comment]
> > You'll need to import the Font Awesome CSS explicitly in `_app.js`: `import '@fortawesome/fontawesome-svg-core/styles.css';`
> 
> Thank you! I did NOT see that in the [Font Awesome React Documentation](https://fontawesome.com/how-to-use/on-the-web/using-with/react), but it's made the difference for me. `<link rel="shortcut icon" href="…" />` in `_app.js` had previously been breaking FontAwesome, but it's all good now.

Hey @markitics - yes, it does fix the problem, but it's still incorrect. Basically, you're importing *all* FontAwesome CSS which shouldn't be needed for FontAwesome React, which is why you aren't seeing it in the documentation. If you look at your bundle size, it's going to become much, much bigger when you import all FontAwesome CSS. It solves the issue because it forces all FontAwesome styles to be imported. Font Awesome React solves this by importing styles needed for the icons you are using, but the bug that I'm describing is breaking that functionality. What you're doing is a brute-force workaround that happens to cover up for the bug (at the expense of increasing the bundle size).

This is still a bug that needs to be tended to!

**Comment by markitics** [comment]
Hi @philiparvidsson, thanks for writing me back – I had overlooked that important issue about dramatically increasing the bundle size. That's no good.

For my specific case of the favicon, I found a different workaround instead: [add a `rewrite`](https://nextjs.org/docs/api-reference/next.config.js/rewrites) in next.config.js

```javascript
// next.config.js
module.exports = {
  images: {
    domains: [... ],
  },
  async rewrites() {
    return [
      {
        source: "/favicon.ico",
        destination: "https://stripe.com/favicon.ico", 
        // URL to .ico on my own CDN (Cloudfront), separate to the app
      },
    ];
  },
  async redirects() {
    return [...];
  },
};

```

It seems the `/favicon.ico` path is a default path that browsers will look for if not directed otherwise. Testing in Brave on desktop, this [`rewrite`](https://nextjs.org/docs/api-reference/next.config.js/rewrites) works perfectly for avoiding the "/favicon.ico not found" error.

But I guess this doesn't really help you add other `<link …` entries, like
* `<link rel="apple-touch-icon"…`
* `<link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32">`
* `<link rel="manifest" href="/manifest.json">`
* etc.

**Comment by wanjas** [comment]
I've tried the example on different versions. 
Looks like `10.0.2-canary-8` introduced this issue with this change https://github.com/vercel/next.js/pull/16758

Strangely enough, I bumped into this issue only when updated my project from 10.0.4 to 1.0.5.

**Comment by blainegarrett** [comment]
Unsure if this is related, but I am getting an empty `<link/>` right before the `<meta name="next-head-count" content="24"/>`. It isn't breaking anything for me, but shows as an error in the schema validation with error: 'Error: Element link is missing one or more of the following attributes: href, itemprop, property, rel, resource.'  Using next 11.0.0

**Comment by volperoid** [comment]
@philiparvidsson Thanks for pointing out this issue; you're a real god-send. I'd probably end up grinding my teeth against the office wall trying to force my icons into submission alongside much larger issues - like async hell. I too noticed the CSS inline classes disappearing but I had no idea what was causing it. I think a possible solution may be to import a favicon using `pages/_document.js`, as it does not rely on the next/head module but injects a <head> element directly into the HTML DOM. Either that or be at the mercy of browser defaults and just add a favicon.ico file into your root, granted you aren't stuck with a .png like I am. Do tell if you find a solution in the mean time. Cheers!

**React works in strange ways...**

**Comment by theDanielJLewis** [comment]
More than a year later, and this is still a bug? Or has there been a solution that doesn't bloat the bundle?

**Comment by allancaffee** [comment]
did a little digging on this and wanted to share what I've learned.

This issue persists for me in Next `12.3.2`. Even adding a `<title>foo</title>` tag in the header causes the inline `style` tag that `react-fontawesome` adds to be removed. Using the workaround of [pulling the whole set of icons in _app.tsx](https://github.com/FortAwesome/react-fontawesome/blob/master/examples/nextjs-app/pages/_app.js) works, but naturally increases the bundle size, wiping out some of the advantages of `react-fontawesome`.

The crux of the problem is definitely that the underlying header code is removing and replacing the incorrect elements. Which is a duplicate of #11012. It looks like a fix for this was merged with #16758 back in 2020, but then reverted not long after with #18713. It seems that approach broke  the use of React components within the `Head` component, but based on discussion within that issue its not clear whether that was ever intended to work.

Is anyone working on a long-term fix to help Next correctly identify its own head elements? This seems likely to break lots of integrations that depend on content dynamically added to head.
